File: src/engine/handEvaluator.js
// Pure 7-card Texas Hold'em hand evaluator. No React.
// Ported from personal-poker-trainer/frontend/src/engine/handEvaluator.ts.
// Returns { handRank, kickers } — no description text (we don't display it).

export const HandRank = {
  HIGH_CARD: 0,
  PAIR: 1,
  TWO_PAIR: 2,
  THREE_OF_A_KIND: 3,
  STRAIGHT: 4,
  FLUSH: 5,
  FULL_HOUSE: 6,
  FOUR_OF_A_KIND: 7,
  STRAIGHT_FLUSH: 8,
  ROYAL_FLUSH: 9,
}

export function evaluateHand(sevenCards) {
  let bestRank = null
  let bestKickers = null

  for (const combo of combinations(sevenCards, 5)) {
    const { handRank, kickers } = evaluateFive(combo)
    if (
      bestRank === null ||
      compareHands(handRank, kickers, bestRank, bestKickers) > 0
    ) {
      bestRank = handRank
      bestKickers = kickers
    }
  }

  return { handRank: bestRank, kickers: bestKickers }
}

export function compareHandResults(a, b) {
  return compareHands(a.handRank, a.kickers, b.handRank, b.kickers)
}

function compareHands(rankA, kickersA, rankB, kickersB) {
  if (rankA !== rankB) return rankA - rankB
  for (let i = 0; i < Math.min(kickersA.length, kickersB.length); i++) {
    if (kickersA[i] !== kickersB[i]) return kickersA[i] - kickersB[i]
  }
  return 0
}

function evaluateFive(cards) {
  const ranks = cards.map((c) => c.rank).sort((a, b) => b - a)
  const suits = cards.map((c) => c.suit)
  const isFlush = new Set(suits).size === 1
  const { isStraight, highCard } = checkStraight(ranks)

  const rankCounts = new Map()
  for (const r of ranks) {
    rankCounts.set(r, (rankCounts.get(r) || 0) + 1)
  }

  const groups = Array.from(rankCounts.entries()).sort(
    (a, b) => b[1] - a[1] || b[0] - a[0],
  )
  const counts = groups.map((g) => g[1])

  if (isStraight && isFlush) {
    if (highCard === 14) {
      return { handRank: HandRank.ROYAL_FLUSH, kickers: [14] }
    }
    return { handRank: HandRank.STRAIGHT_FLUSH, kickers: [highCard] }
  }

  if (counts[0] === 4) {
    return {
      handRank: HandRank.FOUR_OF_A_KIND,
      kickers: [groups[0][0], groups[1][0]],
    }
  }

  if (counts[0] === 3 && counts[1] === 2) {
    return {
      handRank: HandRank.FULL_HOUSE,
      kickers: [groups[0][0], groups[1][0]],
    }
  }

  if (isFlush) {
    return { handRank: HandRank.FLUSH, kickers: ranks }
  }

  if (isStraight) {
    return { handRank: HandRank.STRAIGHT, kickers: [highCard] }
  }

  if (counts[0] === 3) {
    const kickers = groups.slice(1).map((g) => g[0]).sort((a, b) => b - a)
    return {
      handRank: HandRank.THREE_OF_A_KIND,
      kickers: [groups[0][0], ...kickers],
    }
  }

  if (counts[0] === 2 && counts[1] === 2) {
    const pairs = [groups[0][0], groups[1][0]].sort((a, b) => b - a)
    return {
      handRank: HandRank.TWO_PAIR,
      kickers: [...pairs, groups[2][0]],
    }
  }

  if (counts[0] === 2) {
    const kickers = groups.slice(1).map((g) => g[0]).sort((a, b) => b - a)
    return {
      handRank: HandRank.PAIR,
      kickers: [groups[0][0], ...kickers],
    }
  }

  return { handRank: HandRank.HIGH_CARD, kickers: ranks }
}

function checkStraight(sortedRanks) {
  const unique = [...new Set(sortedRanks)].sort((a, b) => b - a)
  if (unique.length !== 5) return { isStraight: false, highCard: 0 }
  if (unique[0] - unique[4] === 4) {
    return { isStraight: true, highCard: unique[0] }
  }
  // Wheel: A-2-3-4-5
  if (
    unique[0] === 14 &&
    unique[1] === 5 &&
    unique[2] === 4 &&
    unique[3] === 3 &&
    unique[4] === 2
  ) {
    return { isStraight: true, highCard: 5 }
  }
  return { isStraight: false, highCard: 0 }
}

function* combinations(arr, k) {
  const n = arr.length
  if (k > n) return
  const idx = Array.from({ length: k }, (_, i) => i)
  while (true) {
    yield idx.map((i) => arr[i])
    let i = k - 1
    while (i >= 0 && idx[i] === n - k + i) i--
    if (i < 0) return
    idx[i]++
    for (let j = i + 1; j < k; j++) idx[j] = idx[j - 1] + 1
  }
}

File: training/games/hand_evaluator.py
"""Pure 7-card Texas Hold'em hand evaluator.

Port of ../src/engine/handEvaluator.js. Returns dicts with `handRank` and
`kickers` — same shape as the JS side so parity checks compare byte-for-byte.
Cards are dicts: {'rank': int 2..14, 'suit': str 'h'|'d'|'c'|'s'}.
"""

from __future__ import annotations

from itertools import combinations


HAND_RANK = {
    "HIGH_CARD": 0,
    "PAIR": 1,
    "TWO_PAIR": 2,
    "THREE_OF_A_KIND": 3,
    "STRAIGHT": 4,
    "FLUSH": 5,
    "FULL_HOUSE": 6,
    "FOUR_OF_A_KIND": 7,
    "STRAIGHT_FLUSH": 8,
    "ROYAL_FLUSH": 9,
}


def evaluate_hand(seven_cards):
    best_rank = None
    best_kickers = None
    for combo in combinations(seven_cards, 5):
        rank, kickers = _evaluate_five(combo)
        if best_rank is None or _compare(rank, kickers, best_rank, best_kickers) > 0:
            best_rank = rank
            best_kickers = kickers
    return {"handRank": best_rank, "kickers": best_kickers}


def compare_hand_results(a, b):
    return _compare(a["handRank"], a["kickers"], b["handRank"], b["kickers"])


def _compare(rank_a, kickers_a, rank_b, kickers_b):
    if rank_a != rank_b:
        return rank_a - rank_b
    for i in range(min(len(kickers_a), len(kickers_b))):
        if kickers_a[i] != kickers_b[i]:
            return kickers_a[i] - kickers_b[i]
    return 0


def _evaluate_five(cards):
    ranks = sorted((c["rank"] for c in cards), reverse=True)
    suits = [c["suit"] for c in cards]
    is_flush = len(set(suits)) == 1
    is_straight, high_card = _check_straight(ranks)

    rank_counts = {}
    for r in ranks:
        rank_counts[r] = rank_counts.get(r, 0) + 1

    groups = sorted(rank_counts.items(), key=lambda g: (-g[1], -g[0]))
    counts = [g[1] for g in groups]

    if is_straight and is_flush:
        if high_card == 14:
            return HAND_RANK["ROYAL_FLUSH"], [14]
        return HAND_RANK["STRAIGHT_FLUSH"], [high_card]

    if counts[0] == 4:
        return HAND_RANK["FOUR_OF_A_KIND"], [groups[0][0], groups[1][0]]

    if counts[0] == 3 and counts[1] == 2:
        return HAND_RANK["FULL_HOUSE"], [groups[0][0], groups[1][0]]

    if is_flush:
        return HAND_RANK["FLUSH"], list(ranks)

    if is_straight:
        return HAND_RANK["STRAIGHT"], [high_card]

    if counts[0] == 3:
        kickers = sorted((g[0] for g in groups[1:]), reverse=True)
        return HAND_RANK["THREE_OF_A_KIND"], [groups[0][0], *kickers]

    if counts[0] == 2 and counts[1] == 2:
        pairs = sorted([groups[0][0], groups[1][0]], reverse=True)
        return HAND_RANK["TWO_PAIR"], [*pairs, groups[2][0]]

    if counts[0] == 2:
        kickers = sorted((g[0] for g in groups[1:]), reverse=True)
        return HAND_RANK["PAIR"], [groups[0][0], *kickers]

    return HAND_RANK["HIGH_CARD"], list(ranks)


def _check_straight(sorted_ranks):
    unique = sorted(set(sorted_ranks), reverse=True)
    if len(unique) != 5:
        return False, 0
    if unique[0] - unique[4] == 4:
        return True, unique[0]
    # Wheel: A-2-3-4-5
    if unique == [14, 5, 4, 3, 2]:
        return True, 5
    return False, 0
